crates/esplora/src/blocking_ext.rs

```diff
@@ -68,250 +56,368 @@ pub trait EsploraExt {
-                                },
-                            ))
-                        });
-
-                        for (outpoint, txout) in previous_outputs {
-                            let _ = graph.insert_txout(outpoint, txout);
-                        }
-                    }
-                }
+        let fetched_hash = match fetched_blocks.entry(height) {
+            btree_map::Entry::Occupied(entry) => *entry.get(),
+            btree_map::Entry::Vacant(entry) => *entry.insert(client.get_block_hash(height)?),
+        };
 
-                let last_index = last_index.expect("Must be set since handles wasn't empty.");
-                let past_gap_limit = if let Some(i) = last_active_index {
-                    last_index > i.saturating_add(stop_gap as u32)
-                } else {
-                    last_index >= stop_gap as u32
-                };
-                if past_gap_limit {
-                    break;
-                }
+        // We have found point of agreement so the update will connect!
+        if fetched_hash == local_hash {
+            break;
+        }
+    }
+
+    Ok(fetched_blocks)
+}
+
+/// Fetches missing checkpoints and finalizes the [`local_chain::Update`].
+///
+/// A checkpoint is considered "missing" if an anchor (of `anchors`) points to a height without an
+/// existing checkpoint/block under `local_tip` or `update_blocks`.
+#[doc(hidden)]
+pub fn finalize_chain_update_blocking<A: Anchor>(
+    client: &esplora_client::BlockingClient,
+    local_tip: &CheckPoint,
+    anchors: &BTreeSet<(A, Txid)>,
+    mut update_blocks: BTreeMap<u32, BlockHash>,
+) -> Result<local_chain::Update, Error> {
+    let update_tip_height = update_blocks
+        .keys()
+        .last()
+        .copied()
+        .expect("must atleast have one block");
+
+    // We want to have a corresponding checkpoint per height. We iterate the heights of anchors
+    // backwards, comparing it against our `local_tip`'s chain and our current set of
+    // `update_blocks` to see if a corresponding checkpoint already exists.
+    let anchor_heights = anchors
+        .iter()
+        .map(|(a, _)| a.anchor_block().height)
+        // filter out duplicate heights
+        .filter({
+            let mut prev_height = Option::<u32>::None;
+            move |h| match prev_height.replace(*h) {
+                None => true,
+                Some(prev_h) => prev_h != *h,
             }
+        })
+        // filter out heights that surpass the update tip
+        .filter(|h| *h <= update_tip_height)
+        .rev();
+
+    // We keep track of a checkpoint node of `local_tip` to make traversing the linked-list of
+    // checkpoints more efficient.
+    let mut curr_cp = local_tip.clone();
 
-            if let Some(last_active_index) = last_active_index {
-                last_active_indexes.insert(keychain, last_active_index);
+    for h in anchor_heights {
+        if let Some(cp) = curr_cp.query_from(h) {
```

Comment:
Don't I just want to know if we have the hash at `h`. Why use `query_from` rather than `query`.

Also it looks like `query_from` should just be generalized into a `range(..)` type method to have more idiomatic and easy to understand name (if it needs to exist right now).

Reply:
We are using `.query_from` as an optimization; `.query_from(height)` returns either the checkpoint of the given `height`, or the one just greater than `height`. Either way, we store the returned `CheckPoint` in `curr_cp` which we can call `.query_from` on again.

Just for some context, I originally wanted to create an update by extending from `curr_cp` (but I deemed it too complicated).

Reply:
@LLFourn I think the `range`-type methods usually return a list of items (while we want this method to return one item). I think a better name would be `query_until` (with the same function signature)...

Reply:
If you want what `query_from` does you can do `query(..).first()` :)

Reply:
Updated #1369 to have `CheckPoint::range`. However, I still need to do tests.